gridstatus/pjm.py

```diff
@@ -538,6 +603,447 @@ def _get_key(self):
+                checkbox_name = checkbox_elem.get("name")
+                matches = re.search(r":([0-9]+):(j_.+)$", checkbox_name)
+                if matches is not None:
+                    to_check_idx = int(matches.group(1))
+                    if source_id is None:
+                        source_id = re.sub(r"_input$", "", matches.group(2))
+                checkboxes[to_check_idx] = checked
+
+        return (
+            checkboxes,
+            source_id,
+        )
+
+    def _dv_lmp_select_checkbox(
+        self,
+        dv_session,
+        form_source_id,
+        checkboxes,
+        to_check_idx,
+        verbose=False,
+    ):
+        # toggle checkbox
+        params = {
+            "javax.faces.partial.ajax": "true",
+            "javax.faces.source": f"chart1FrmLmpSelection:tblBusAggregates:{to_check_idx}:{form_source_id}",  # noqa: E501
+            "javax.faces.partial.execute": f"chart1FrmLmpSelection:tblBusAggregates:{to_check_idx}:{form_source_id}",  # noqa: E501
+            "javax.faces.partial.render": "globalMessages",
+            "javax.faces.behavior.event": "change",
+            "javax.faces.partial.event": "change",
+            "chart1FrmLmpSelection": "chart1FrmLmpSelection",
+            "chart1FrmLmpSelection:tblBusAggregates:name:filter": "",
+            "chart1FrmLmpSelection:tblBusAggregates:voltage:filter": "",
+            "chart1FrmLmpSelection:tblBusAggregates:station:filter": "",
+            "chart1FrmLmpSelection:tblBusAggregates_scrollState": "0,0",
+        }
+        checkboxes[to_check_idx] = True
+        for idx, is_checked in checkboxes.items():
+            if is_checked:
+                params[
+                    f"chart1FrmLmpSelection:tblBusAggregates"
+                    f":{idx}:{form_source_id}_input"  # noqa: E501
+                ] = "on"
+        response = self._dv_lmp_fetch(
+            dv_session,
+            params,
+            verbose=verbose,
+        )
+        if verbose:
+            print(f"response (truncated) = {response.content[0:256]}", file=sys.stderr)
+
+    def _get_lmp_via_dv(
+        self,
+        date,
+        market: str,
+        end=None,
+        locations="hubs",
+        location_type=None,
+        verbose=False,
+    ):
+        """Get latest LMP data from Data Viewer, which includes RT & DA"""
+        with requests.session() as session:
+            initial_fetch = session.get(DATAVIEWER_LMP_URL)
+
+            dv_session = self._new_dv_session(session, verbose=verbose)
+            if dv_session is None:
+                raise ValueError("Could not create new DV Session")
+
+            chart_ids = self._dv_lmp_extract_chart_ids(initial_fetch, verbose=verbose)
+            if chart_ids is None:
+                raise ValueError("Could not get LMP Chart IDs")
+            dv_session.update(chart_ids)
+
+            df = self._dv_lmp_fetch_data(dv_session, verbose=verbose)
```

Comment:
would it make more sense to move the calls to `_new_dv_session` and `_dv_lmp_extract_chart_ids` to inside `_dv_lmp_fetch_data` method? 

you cant call `_dv_lmp_fetch_data` if you haven't already called  `_new_dv_session` / `_dv_lmp_extract_chart_ids` and it's not like we are using the return values from those method outside this method. so, not sure what the advantage to having them outside is and again it would make this function simpler.

Reply:
Done